src/java/org/apache/cassandra/journal/Journal.java

```diff
@@ -112,31 +116,92 @@ public class Journal<K, V> implements Shutdownable
     private final WaitQueue segmentPrepared = newWaitQueue();
     private final WaitQueue allocatorThreadWaitQueue = newWaitQueue();
     private final BooleanSupplier allocatorThreadWaitCondition = () -> (availableSegment == null);
+    private final FlusherCallbacks flusherCallbacks;
 
     SequentialExecutorPlus closer;
     //private final Set<Descriptor> invalidations = Collections.newSetFromMap(new ConcurrentHashMap<>());
 
+    private class FlusherCallbacks implements Flusher.Callbacks
+    {
+        private final Queue<WaitingFor> waitingFor = new ConcurrentLinkedQueue<>();
+
+        @Override
+        public void onFlush(long segment, int position)
+        {
+            Iterator<WaitingFor> iter = waitingFor.iterator();
+            while (iter.hasNext())
+            {
+                WaitingFor wait = iter.next();
+                if (wait.segment == segment && wait.position <= position)
+                {
+                    wait.run();
+                    iter.remove();
+                }
+            }
+        }
+
+        @Override
+        public void onFlushFailed(Throwable cause)
+        {
+            // TODO: panic
+        }
+
+        public void submit(RecordPointer pointer, Runnable runnable)
+        {
+            WaitingFor wait = new WaitingFor(pointer.segment, pointer.position, runnable);
+            waitingFor.add(wait);
+            if (isFlushed(pointer))
+                wait.run();
+        }
+    }
+
+    private static class WaitingFor extends RecordPointer implements Runnable
+    {
+        private final AtomicReference<Runnable> onFlushed;
```

Comment:
It would be nice not to rely on additional atomics here. Is there any reason the journal thread won't promptly catch up any that could have run immediately?

Reply:
It will catch up. Atomic's purpose was to avoid double-running, but it should never happen, just a precaution really.

Reply:
I mean if we don't invoke on the caller that submits - I agree this can permit double invocation, but we could just not do this?

Reply:
Yeah; the intention here was to make sure that we run the runnable only once. You are of course right that it journal thread will promptly catch up any that could have run immediately.